File: 2019/day17/cpp/src/asci.cpp
#include "asci.hpp"
#include <math.h>
#include <boost/algorithm/string.hpp>
#include <boost/algorithm/string/replace.hpp>
#include "ncurses.h"
#include <signal.h>
#include <iostream>

Asci::Asci(std::string file_name) : 
    m_file_name(file_name), m_intcode(), m_mp() {

  m_intcode = Intcode(m_file_name,{},false,false);
  m_mp = get_map();

}

std::map<std::pair<int,int>,int> Asci::get_map() {

  // Initial position
  int y(0), x(0);

  std::map<std::pair<int,int>,int> map;

  // Go through the intcode output
  while (true) {
    auto output = m_intcode();
    if (std::get<0>(output)) { 
      break;
    }
    if (std::get<1>(output) == 10) {
      y += 1;
      x = 0;
    } else {
      map[std::make_pair(y, x)] = std::get<1>(output);
      x += 1;
    }
  }
  return map;
}

int Asci::compute_intersections(bool print_map_) {

  int tot(0);
  for (auto &m : m_mp) {
    // If the value is 35 ('#') check top, left, right, bottom
    if (m.second == 35) {
      int y = m.first.first, x = m.first.second;
      if (m_mp[std::make_pair(y, x + 1)] == 35 &&
          m_mp[std::make_pair(y, x - 1)] == 35 &&
          m_mp[std::make_pair(y - 1, x)] == 35 &&
          m_mp[std::make_pair(y + 1, x)] == 35) {
        m_mp[std::make_pair(y,x)] = int('O');
        tot += y*x;
      }
    }
  }
  // Print the map
  if (print_map_) {
    print_map();
  }
  return tot;
}

void Asci::print_map() {

  int max_x(WINT_MIN);
  int max_y(WINT_MIN);

  for (auto &m : m_mp) {
    max_y = (max_y < m.first.first) ? m.first.first : max_y;
    max_x = (max_x < m.first.second) ? m.first.second : max_x;
  }

  for (int row = 0; row < max_y+1;row++) {
    for (int col = 0; col < max_x+1; col++) {
      std::cout << char(m_mp[std::make_pair(row,col)]);
    }
    std::cout<<std::endl;
  }
}

void Asci::find_start_pose(int &x, int &y, double &theta) {
  // Look foor (<,v,^,>)
  for (auto &m : m_mp) {
    if (m.second == 94 || m.second == 60 || m.second == 62 || m.second == 118) {
      y = m.first.first;
      x = m.first.second;
      break;
    }
  }
  // Initial point - figure out orientation
  auto v = m_mp[std::make_pair(y,x)];
  if (v == int('^')) {
    theta = M_PI_2;
  } else if (v == int('<')) {
    theta = -M_PI;
  } else if (v == int('>')) {
    theta = 0;
  } else {
    theta = -M_PI_2;
  }
}

std::string Asci::traverse_scaffold() {
  // Starting position
  int x,y;
  double theta,theta_;
  find_start_pose(x,y,theta);
  double dy=(-sin(theta)), dx=(cos(theta));

  // Keep track of how many straight steps
  int straight(0); 
  std::string instruction_string = "";

  // Keep going until we reach the end
  while (true) {

    std::pair<int,int> pos(y+dy,x+dx);
    // If scaffold infront of us, keep going
    if (m_mp[pos] == int('#') || m_mp[pos] == int('O')) {
      m_mp[std::make_pair(y,x)] = int('O');
      straight++;
      y = y+dy;
      x = x+dx;  
    }
    // Otherwise, see where the scaffold turns
    else {
      if      (m_mp[std::make_pair(y-1,x)] == int('#')) { theta_ = M_PI_2;}
      else if (m_mp[std::make_pair(y+1,x)] == int('#')) { theta_ = -M_PI_2;}
      else if (m_mp[std::make_pair(y,x-1)] == int('#')) { theta_ = -M_PI;}
      else if (m_mp[std::make_pair(y,x+1)] == int('#')) { theta_ = 0;}
      // Reached the end
      else {
        m_mp[std::make_pair(y,x)] = int('x');
        if (straight > 0) {
          instruction_string += std::to_string(straight);
        }
        return instruction_string;
      }

      // Apply the steps taken and reset <straight>
      if (straight > 0) {
        instruction_string += std::to_string(straight)+",";
        straight = 0;
      }

      // Find angle between theta and theta_
      double d = sin(theta+M_PI_2)*sin(theta_) + cos(theta+M_PI_2)*cos(theta_);
      d = (d>1) ? 1 : ((d<-1) ? -1 : d);

      if (abs(acos(d)) < 1e-4) {
        instruction_string += "L,";
        theta = theta+M_PI_2;
      } else {
        instruction_string += "R,";
        theta = theta - M_PI_2;
      }

      // Update directions      
      dy = int(round(-sin(theta)));
      dx = int(round(cos(theta)));
    }
  }
  return "Solution failed";
}

void Asci::find_movement_routine(
    std::string &routine, std::string &A,std::string &B,std::string &C) {

  // Get the instruction string
  std::string str0 = traverse_scaffold()+",";

  // Chop up the string from 20 characters to 1 characters for A
  for (size_t an = std::min(size_t(20),str0.size()); an > 0; an--) {
    // Don't include the comma
    if (str0.substr(0,an).back() == ',') {
      continue;
    }
    // Otherwise, remove all such occurences, and do same thing for B
    else {
      std::string str1 = str0;
      boost::erase_all(str1, str0.substr(0,an+1));

      // Chop up the string from 20 characters to 1 characters for B
      for (size_t bn = std::min(size_t(20),str1.size()); bn > 0; bn--) {
        // Don't include the comma
        if (str1.substr(0,bn).back() == ',') {
          continue;
        // Otherwise, remove all such occurences, and do same thing for C
        } else {
          auto str2 = str1;
          boost::erase_all(str2, str1.substr(0, bn + 1));
          // Chop up the string from 20 characters to 1 characters for C
          for (size_t cn = std::min(size_t(20), str2.size()); cn > 0; cn--) {
            // Don't include the comma
            if (str2.substr(0,cn).back() == ',') {
              continue;
              // Otherwise, remove all such occurences
            } else {
              auto str3 = str2;
              boost::erase_all(str3, str2.substr(0, cn + 1));

              // If remainder is empty, we are done
              if (str3.size() == 0 ) {
                // Build the routine
                A = str0.substr(0, an);
                B = str1.substr(0,bn);
                C = str2.substr(0,cn);
                boost::replace_all(str0, A, "A");
                boost::replace_all(str0, B, "B");
                boost::replace_all(str0, C, "C");
                str0.pop_back();
                routine = str0;
                return;
              }
            }
          }
        }
      }
    }
  }
}

std::vector<int64_t> Asci::construct_input(bool print_iterations) {
  std::string routine, A, B, C;
  find_movement_routine(routine,A,B,C);
  std::vector<int64_t> inp;
  for (auto &r : routine) { inp.push_back(int(r)); }
  inp.push_back(10);
  
  for (auto &r : A) { inp.push_back(int(r)); }
  inp.push_back(10);
  
  for (auto &r : B) { inp.push_back(int(r)); }
  inp.push_back(10);
  
  for (auto &r : C) { inp.push_back(int(r)); }
  inp.push_back(10);

  if(print_iterations) { inp.push_back(int('y')); } 
  else { inp.push_back(int('n')); }
  inp.push_back(10);

  return inp;
}


int Asci::collect_dust(bool print_iterations) {

  // Pass the intput to the intcode computer
  auto input = construct_input(print_iterations);
  m_intcode = Intcode(m_file_name,input,false,false);
  m_intcode.m_program[0] = 2;

  // Run until it terminates
  if(!print_iterations) {
    while(true) {
      auto output = m_intcode();
      if (std::get<0>(output)) {
        return std::get<1>(output);
      }
    }
  } else {
    return print_video_feed();
  }
}


int Asci::print_video_feed() {

  signal(SIGINT, catch_ctrlc);
  initscr();

  int height, width;
  getmaxyx(stdscr, height, width);

  std::string valid = "#.<^v>\n";
  int max_x(WINT_MIN);
  int max_y(WINT_MIN);

  for (auto &m : m_mp) {
    max_y = (max_y < m.first.first) ? m.first.first : max_y;
    max_x = (max_x < m.first.second) ? m.first.second : max_x;
  }

  std::map<std::pair<int,int>,int> map;
  while (true) {
    int y(0),x(0);
    bool skip(false);
    // Run until termination
    while (true) {
      auto output = m_intcode();
      int o = std::get<1>(output);
      if (std::get<0>(output)) {
        endwin();
        return o;}

      if (valid.find(char(o % 256)) == std::string::npos ||
          (x == 0 && o == 10)) {
        skip = true;
      }
      // char(10) == '\n' new line
      if (o == 10) {
        if(!skip) { y+=1; }
        x = 0; skip = false;
        if (y > max_y-1) { break; }
      // else, fill the map
      } else if (!skip) {
        map[std::make_pair(y,x)] = o;
        x += 1;
      }
    }
    // Print the map
    for (int row = 0; row < max_y+1; row++) {
      std::string msg = "";
      for (int col = 0; col < max_x+1; col++) {
        if (col > width -3) {break;}
        msg += char(map[std::make_pair(row,col)]);
      }
      if (row > height -3) { break;}
      mvprintw(row, 0, msg.c_str());
    }
    refresh();
  }
  return -1;
}

void Asci::catch_ctrlc(int s)
{
  endwin();
  printf("Caught signal %d\n", s);
  exit(1);
}

File: 2019/day17/python/asci.py
from os.path import dirname
from os.path import realpath
from os.path import join
from collections import defaultdict

import curses
import numpy as np

import sys

sys.path.append(join(dirname(realpath(__file__)), *["..", ".."]))
from day5.python.intcode import Intcode


class Asci:
    """ Class that controls the scaffolding bot. """

    def __init__(self, fileName):
        """ Initialize class and intcode computer. """
        self.fileName = fileName
        self.ic = Intcode(fileName, verbose=False, reset=False)
        self.mp = self.get_map()

    def get_map(self):
        """ Call the intcode computer to generate the map """

        self.mp = defaultdict(lambda: ord("x"))
        y, x = 0, 0
        while True:
            cond, output = self.ic()

            if cond:
                break
            # New row of the print out
            if output == 10:
                y += 1
                x = 0
            # Assign the value to the map
            else:
                self.mp[y, x] = output
                x += 1

        return self.mp

    def compute_intersections(self):
        """ Go through map and find if any section intersects """

        # Make list of elements where v ==35 (chr(v) = '#')
        list = [k for k, v in self.mp.items() if v == 35]
        sum = 0
        for y, x in list:
            x1, x2 = self.mp[y, x + 1], self.mp[y, x - 1]
            y1, y2 = self.mp[y - 1, x], self.mp[y + 1, x]
            if x1 == 35 and x2 == 35 and y1 == 35 and y2 == 35:
                self.mp[y, x] = ord("O")
                sum += y * x
        return sum

    def find_start_pose(self):
        """ Find where (<,>,^,v) is in the map """

        # Find start position
        y, x = [
            k
            for k, v in self.mp.items()
            if v == 94 or v == 60 or v == 62 or v == 118
        ][0]

        # Assign orientation
        theta = 0
        if self.mp[y, x] == ord("^"):
            theta = np.pi / 2
        elif self.mp[y, x] == ord("<"):
            theta = -np.pi
        elif self.mp[y, x] == ord(">"):
            theta = 0
        else:
            theta = -np.pi / 2

        return y, x, theta

    def traverse_scaffold(self):
        """ Compute which actions are needed to go through scaffold """

        y, x, theta = self.find_start_pose()
        dy, dx = -np.sin(theta), np.cos(theta)
        straight, instruction_string = 0, ""

        while True:
            # Either scaffold infront of us
            if self.mp[y + dy, x + dx] == ord("#") or self.mp[
                y + dy, x + dx
            ] == ord("O"):
                self.mp[y, x] = ord("O")
                straight += 1
                y, x = y + dy, x + dx
            # Otherwise, it is to our left or right
            else:
                if self.mp[y - 1, x] == ord("#"):
                    theta_ = np.pi / 2
                elif self.mp[y + 1, x] == ord("#"):
                    theta_ = -np.pi / 2
                elif self.mp[y, x - 1] == ord("#"):
                    theta_ = -np.pi
                elif self.mp[y, x + 1] == ord("#"):
                    theta_ = 0

                # Reached the end
                else:
                    self.mp[y, x] = ord("x")
                    if straight > 0:
                        instruction_string = instruction_string + str(straight)

                    return instruction_string

                if straight > 0:
                    instruction_string = (
                        instruction_string + str(straight) + ","
                    )
                    straight = 0

                # Find angle between directions
                d = np.arccos(
                    np.dot(
                        [np.sin(theta + np.pi / 2), np.cos(theta + np.pi / 2)],
                        [np.sin(theta_), np.cos(theta_)],
                    )
                )

                if d == 0:
                    instruction_string += "L,"
                    theta = theta + np.pi / 2
                else:
                    instruction_string += "R,"
                    theta = theta - np.pi / 2
                # Update orientation

                dy, dx = np.int(-np.sin(theta)), np.int(np.cos(theta))

    def find_movement_routine(self):
        """ Find three repeating patterns in the movement string """

        str0 = self.traverse_scaffold() + ","

        for an in range(min(20, len(str0)), 0, -1):
            if str0[:an][-1] == ",":
                continue
            else:
                str1 = str0.replace(str0[0 : an + 1], "")
                for bn in range(min(20, len(str1)), 0, -1):
                    if str1[:bn][-1] == ",":
                        continue
                    else:
                        str2 = str1.replace(str1[0 : bn + 1], "")
                        for cn in range(min(20, len(str2)), 0, -1):
                            if str2[:cn][-1] == ",":
                                continue
                            else:
                                str3 = str2.replace(str2[0 : cn + 1], "")
                                if not str3:
                                    A, B, C = (
                                        str0[0:an],
                                        str1[0:bn],
                                        str2[0:cn],
                                    )
                                    routine = (
                                        str0.replace(A, "A")
                                        .replace(B, "B")
                                        .replace(C, "C")[:-1]
                                    )
                                    return routine, A, B, C

    def construct_input(self, print_iterations=False):
        """Get the routine and movement functions and
        translate into the desired input format"""
        routine, A, B, C = self.find_movement_routine()

        def toOrd(A):
            return [ord(x) for x in A]

        output = toOrd(routine) + [10] + toOrd(A) + [10] + toOrd(B) + [10]
        output += toOrd(C) + [10] + [ord("y" if print_iterations else "n"), 10]

        return output

    def print_video_feed(self):
        """ Print the progress of the scaffolding robot """

        screen = curses.initscr()
        height, width = screen.getmaxyx()

        # String template used to remove unecessary plotting
        valid = "#.<^v>\n"

        y_max, x_max = map(max, zip(*self.mp.keys()))
        mp = defaultdict(lambda: ord(" "))
        try:
            # Keep going until we reach the terminating condition
            while True:
                # Initial conditions for each map
                y, x, skip = 0, 0, False
                while True:
                    cond, output = self.ic()
                    if cond:
                        return output

                    # Check if output belongs to the valid string and
                    # filter new lines
                    if (chr(output % 256) not in valid) or (
                        x == 0 and output == 10
                    ):
                        skip = True

                    # If output =
                    if output == 10:
                        if not skip:
                            y += 1
                        x, skip = 0, False
                        if y > y_max - 1:
                            break
                    elif not skip:
                        mp[y, x] = output
                        x += 1

                # Print using screen
                self.print_map(screen, width, height, y_max, x_max, mp)

        finally:
            curses.endwin()

    def print_map(self, screen, width, height, y_max, x_max, mp):
        for row in range(0, y_max):
            msg = []
            for k in range(0, x_max):
                if k > width - 3:
                    break
                msg.append(chr(mp[row, k]))
                if row > height - 3:
                    break
            screen.addstr(row, 0, "".join(msg))
        screen.refresh()

    def collect_dust(self, print_iterations=False):
        """ Let the robot traverse using the compted input """

        input = self.construct_input(print_iterations)

        self.ic = Intcode(
            self.fileName, input=input, verbose=False, reset=False
        )
        self.ic.memory[0] = 2

        if not print_iterations:
            while True:
                cond, output = self.ic()
                if cond:
                    return output
        else:
            return self.print_video_feed()
